test/test_distributed.py

```diff
@@ -790,37 +801,34 @@ def forward(self, x):
         model_DDP = copy.deepcopy(model)
         model_DDP.cuda(gpu_subset[0])
         model_DDP = nn.parallel.DistributedDataParallel(model_DDP, device_ids=gpu_subset)
-        optimizer = optim.SGD(model_DDP.parameters(), lr=1e-2)
-        optimizer.zero_grad()
 
         # batch_size for DDP should be divisible by #GPU per node.
         batch_size = len(gpu_subset) * int(WORLD_SIZE)
         input_cpu = torch.randn(batch_size, 2)
         target = torch.randn(batch_size, 4)
         loss = nn.MSELoss()
 
-        # single gpu training
-        self._test_DDP_helper(model_gpu,
-                              input_cpu.cuda(gpu_subset[0]),
-                              target.cuda(gpu_subset[0]),
-                              loss)
-
-        # DDP training, DDP scatters subsets of input_cpu to nodes/GPUs
-        self._test_DDP_helper(model_DDP,
-                              input_cpu[rank * len(gpu_subset):(rank + 1) * len(gpu_subset)],
-                              target[rank * len(gpu_subset):(rank + 1) * len(gpu_subset)].cuda(gpu_subset[0]),
-                              loss)
-
-        for layer_gpu, layer_DDP in zip(model_gpu.modules(), model_DDP.module.modules()):
-            if isinstance(layer_gpu, nn.Linear):
-                self.assertEqual(layer_gpu.weight.grad, layer_DDP.weight.grad)
-
-        # Run SGD and second iteration to shake out errors
-        optimizer.step()
-        self._test_DDP_helper(model_DDP,
-                              input_cpu,
-                              target.cuda(gpu_subset[0]),
-                              loss)
+        for i in range(2):
+            # single gpu training
+            self._test_DDP_helper(model_gpu,
+                                  input_cpu.cuda(gpu_subset[0]),
+                                  target.cuda(gpu_subset[0]),
+                                  loss)
+
+            # DDP training, DDP scatters subsets of input_cpu to nodes/GPUs
+            self._test_DDP_helper(model_DDP,
+                                  input_cpu[rank * len(gpu_subset):(rank + 1) * len(gpu_subset)],
+                                  target[rank * len(gpu_subset):(rank + 1) * len(gpu_subset)].cuda(gpu_subset[0]),
+                                  loss)
+
+            # Update weights and run a second iteration to shake out errors
+            model_step(model_gpu)
+            model_step(model_DDP.module)
```

Comment:
The change above would also just let you pass `model_DDP` here without taking `.module`